File: python/flexflow/serve/models/starcoder.py
# Copyright 2023 CMU, Facebook, LANL, MIT, NVIDIA, and Stanford (alphabetical)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from flexflow.core import *
from .base import FlexFlowModel
import random, torch


class STARCODERConfig:
    def __init__(self, hf_config):
        self.max_beam_width = 1
        self.max_beam_depth = 8
        self.max_spec_tree_token_num = 20
        self.dropout_p = hf_config.attn_pdrop
        self.hidden_size = hf_config.n_embd
        self.layer_norm_epsilon = hf_config.layer_norm_epsilon
        self.max_position_embeddings = hf_config.n_positions
        self.num_hidden_layers = hf_config.n_layer
        self.vocab_size = hf_config.vocab_size
        self.intermediate_size = hf_config.n_inner
        self.n_head_kv = 1 if hf_config.multi_query else hf_config.n_head
        self.rotary_embedding_meta = RotaryEmbeddingMeta(apply_rotary_embedding=False)
        # Standardized FlexFlow num heads fields below
        self.num_attention_heads = hf_config.n_head
        self.num_key_value_heads = self.n_head_kv


class FlexFlowSTARCODER(FlexFlowModel):
    def __init__(
        self,
        mode,
        generation_config,
        ffconfig,
        hf_config,
        data_type,
        max_tokens_per_batch,
        weights_filepath="",
        tokenizer_filepath="",
    ):
        self.mode = mode
        self.generation_config = generation_config
        self.ffconfig = ffconfig
        self.data_type = data_type
        self.starcoder_config = STARCODERConfig(hf_config)
        self.weights_filepath = weights_filepath
        self.tokenizer_filepath = tokenizer_filepath
        self.maxint = 2**31 - 1
        max_verify_tokens_per_batch = (
            max_tokens_per_batch + self.starcoder_config.max_spec_tree_token_num
        )

        # Sanity checks
        if (
            self.starcoder_config.hidden_size
            % self.starcoder_config.num_attention_heads
            != 0
        ):
            raise ValueError(
                f"Hidden size ({self.starcoder_config.hidden_size}) is not divisible by n_head ({self.starcoder_config.num_attention_heads})"
            )

        # Sanity checks
        if (
            self.starcoder_config.num_attention_heads
            < self.ffconfig.tensor_parallelism_degree
            or self.starcoder_config.num_attention_heads
            % self.ffconfig.tensor_parallelism_degree
            != 0
        ):
            raise ValueError(
                f"Number of attention heads ({self.starcoder_config.num_attention_heads}) is smaller, or not divisible by tensor parallelism degree ({self.ffconfig.tensor_parallelism_degree})"
            )

        self.build_model(
            max_tokens_per_batch
            if self.mode == InferenceMode.INC_DECODING_MODE
            else max_verify_tokens_per_batch
        )

    def build_model(self, max_tokens_per_batch):
        ffmodel = FFModel(self.ffconfig)

        tokens_dims = [max_tokens_per_batch, 1]
        input_tensor = ffmodel.create_tensor(tokens_dims, DataType.DT_INT32)
        position_tensor = ffmodel.create_tensor(tokens_dims, DataType.DT_INT32)

        embed_init = UniformInitializer(random.randint(0, self.maxint), 0, 0)
        ffmodel.set_position_offset(0)
        token = ffmodel.embedding(
            input_tensor,
            self.starcoder_config.vocab_size,
            self.starcoder_config.hidden_size,
            AggrMode.AGGR_MODE_NONE,
            self.data_type,
            None,
            embed_init,
            name="wte",
        )
        positional_embedding = ffmodel.embedding(
            position_tensor,
            self.starcoder_config.max_position_embeddings,
            self.starcoder_config.hidden_size,
            AggrMode.AGGR_MODE_NONE,
            self.data_type,
            None,
            embed_init,
            name="wpe",
        )

        axes = [
            0,
        ]

        for i in range(self.starcoder_config.num_hidden_layers):
            ffmodel.set_transformer_layer_id(i)

            hidden_states, ln_1 = ffmodel.residual_layer_norm(
                token if i == 0 else residual,
                positional_embedding if i == 0 else c_proj,
                None,
                False,
                axes,
                True,
                self.starcoder_config.layer_norm_epsilon,
                name=f"layers.{i}.ln_1",
            )

            qkv_proj = ffmodel.dense(
                ln_1,
                3 * self.starcoder_config.hidden_size,
                ActiMode.AC_MODE_NONE,
                False,
                name=f"layers.{i}.self_attn.qkv_proj",
            )

            assert self.mode == InferenceMode.INC_DECODING_MODE
            o_proj = ffmodel.inc_multiquery_self_attention(
                qkv_proj,
                self.starcoder_config.hidden_size,
                self.starcoder_config.num_attention_heads,
                self.starcoder_config.n_head_kv,
                self.starcoder_config.hidden_size
                // self.starcoder_config.num_attention_heads,
                self.starcoder_config.hidden_size
                // self.starcoder_config.num_attention_heads,
                0.0,  # dropout
                True,  # qkv_bias
                False,  # final_bias
                False,  # add_zero_attn
                DataType.DT_NONE,  # data_type
                None,  # kernel initializer
                self.starcoder_config.rotary_embedding_meta,
                name=f"layers.{i}.attn.c_attn",
            )

            mha = ffmodel.dense(
                o_proj,
                self.starcoder_config.hidden_size,
                ActiMode.AC_MODE_NONE,
                False,
                name=f"layers.{i}.self_attn.o_proj"
            )

            residual, l2_norm = ffmodel.residual_layer_norm(
                hidden_states,
                mha,
                None,
                False,
                residual,
                axes,
                True,
                self.starcoder_config.layer_norm_epsilon,
                name=f"layers.{i}.ln_2",
            )

            # mlp

            c_fc = ffmodel.dense(
                l2_norm,
                self.starcoder_config.intermediate_size,
                ActiMode.AC_MODE_NONE,
                True,
                name=f"layers.{i}.mlp.c_fc",
            )
            activation = ffmodel.gelu(c_fc, False)
            c_proj = ffmodel.dense(
                activation,
                self.starcoder_config.hidden_size,
                ActiMode.AC_MODE_NONE,
                True,
                name=f"layers.{i}.mlp.c_proj",
            )

        _, ln_f = ffmodel.residual_layer_norm(
            residual,
            c_proj,
            None,
            False,
            axes,
            True,
            self.starcoder_config.layer_norm_epsilon,
            name=f"ln_f",
        )
        lm_head = ffmodel.dense(
            ln_f,
            self.starcoder_config.vocab_size,
            ActiMode.AC_MODE_NONE,
            False,
            name="lm_head",
        )

        if self.generation_config.do_sample:
            dense = ffmodel.scalar_true_divide(
                lm_head, self.generation_config.temperature, False
            )
            softmax = ffmodel.softmax(dense, -1)
            output = ffmodel.sampling(softmax, self.generation_config.topp)
        else:
            softmax = ffmodel.softmax(lm_head, -1)
            output = ffmodel.argmax(softmax, False)

        self.ffmodel = ffmodel

    def convert_hf_model(model, dst_folder):
        os.makedirs(dst_folder, exist_ok=True)
        for name, params in model.named_parameters():
            name = name.replace("transformer.h", "layers").replace("transformer.", "")
            if "attn.c_attn.weight" in name:
                name_q = name.replace("attn.c_attn", "attn.c_attn.q_proj")
                name_k = name.replace("attn.c_attn", "attn.c_attn.k_proj")
                name_v = name.replace("attn.c_attn", "attn.c_attn.v_proj")
                q, k, v = torch.split(
                    params,
                    [
                        model.config.hidden_size,
                        model.config.hidden_size // model.config.num_attention_heads,
                        model.config.hidden_size // model.config.num_attention_heads,
                    ],
                    0,
                )
                q.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_q))
                k.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_k))
                v.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_v))
            elif "attn.c_attn.bias" in name:
                name_q = name.replace("attn.c_attn", "attn.c_attn.q_proj")
                name_k = name.replace("attn.c_attn", "attn.c_attn.k_proj")
                name_v = name.replace("attn.c_attn", "attn.c_attn.v_proj")
                q, k, v = torch.split(
                    params,
                    [
                        model.config.hidden_size,
                        model.config.hidden_size // model.config.num_attention_heads,
                        model.config.hidden_size // model.config.num_attention_heads,
                    ],
                    0,
                )
                q.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_q))
                k.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_k))
                v.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_v))
            elif "attn.c_proj.bias" in name:
                name = name.replace("attn.c_proj", "attn.c_attn.o_proj")
                params.detach().cpu().numpy().tofile(os.path.join(dst_folder, name))
            elif "attn.c_proj.weight" in name:
                name = name.replace("attn.c_proj", "attn.c_attn.o_proj")
                params.detach().cpu().numpy().tofile(os.path.join(dst_folder, name))
            else:
                params.detach().cpu().numpy().tofile(os.path.join(dst_folder, name))
        model.lm_head.weight.detach().cpu().numpy().tofile(
            os.path.join(dst_folder, "lm_head.weight")
        )

File: inference/models/starcoder.cc
/* Copyright 2023 CMU, Facebook, LANL, MIT, NVIDIA, and Stanford (alphabetical)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "starcoder.h"

namespace FlexFlow {

using namespace Legion;
using json = nlohmann::json;

void STARCODER::create_starcoder_model(
    FFModel &ff,
    std::string const &model_config_file_path,
    std::string const &weight_file_path,
    InferenceMode mode,
    GenerationConfig generationConfig,
    bool use_full_precision) {
  // do not apply cpu offload in beam search model.
  STARCODERConfig startcoder_config(model_config_file_path);
  startcoder_config.print();

  if (ff.config.tensor_parallelism_degree >
          startcoder_config.num_attention_heads ||
      startcoder_config.num_attention_heads %
              ff.config.tensor_parallelism_degree !=
          0) {
    assert(false && "The number of attention heads is smaller, or it is not "
                    "divisible by the tensor parallelism degree");
  }

  std::unordered_map<std::string, Layer *> weights_layers;
  std::vector<int> axes = {0};

  Tensor input;
  Tensor position_input;
  ff.set_position_offset(0);
  {
    // assert(startcoder_config.max_num_tokens <= BatchConfig::MAX_NUM_TOKENS);
    int const token_dims[] = {
        (mode == TREE_VERIFY_MODE || mode == BEAM_SEARCH_MODE)
            ? BatchConfig::max_verify_tokens_per_batch()
            : BatchConfig::max_tokens_per_batch(),
        1};
    input = ff.create_tensor<2>(token_dims, DT_INT32);
    position_input = ff.create_tensor<2>(token_dims, DT_INT32);
  }

  Initializer *embed_init = new UniformInitializer(std::rand(), 0, 0);

  Tensor token = ff.embedding(input,
                              startcoder_config.vocab_size,
                              startcoder_config.hidden_size,
                              AGGR_MODE_NONE,
                              use_full_precision ? DT_FLOAT : DT_HALF,
                              NULL,
                              embed_init,
                              "transformer_wte");

  Tensor positional_embedding =
      ff.embedding(position_input,
                   startcoder_config.max_position_embeddings,
                   startcoder_config.hidden_size,
                   AGGR_MODE_NONE,
                   use_full_precision ? DT_FLOAT : DT_HALF,
                   NULL,
                   embed_init,
                   "transformer_wpe");

  Tensor residual = nullptr, c_proj = nullptr;
  Tensor res_ln_outputs[2] = {nullptr, nullptr};

  for (int i = 0; i < startcoder_config.num_hidden_layers; i++) {
    // set transformer layer id
    ff.set_transformer_layer_id(i);

    // step 1: attention
    ff.residual_layer_norm(
        (i == 0) ? token : residual,
        (i == 0) ? positional_embedding : c_proj,
        nullptr,
        res_ln_outputs,
        false,
        axes,
        true,
        startcoder_config.layer_norm_epsilon,
        true,
        DT_NONE,
        std::string("layers_" + std::to_string(i) + "_ln_1").c_str());
    Tensor hidden_states = res_ln_outputs[0];
    Tensor ln_1 = res_ln_outputs[1];

    Tensor mha;
    switch (mode) {
      case INC_DECODING_MODE: {
        mha = ff.inc_multiquery_self_attention(
            ln_1,
            startcoder_config.hidden_size,
            startcoder_config.num_attention_heads,
            1,
            startcoder_config.hidden_size /
                startcoder_config.num_attention_heads,
            startcoder_config.hidden_size /
                startcoder_config.num_attention_heads,
            startcoder_config.dropout_p, /*dropout*/
            true,                        /*bias*/
            false,                       /*add_bias_kv*/
            false,                       /*add_zero_attn*/
            DT_NONE,                     /*data_type*/
            nullptr,                     /*kernel_initializer*/
            false,                       /*apply_rotary_embedding*/
            false,                       /*scaling query*/
            1.0f,                        /*scaling factor*/
            true,                        /*qk_prod_scaling*/
            false,                       /*position_bias*/
            std::string("layers_" + std::to_string(i) + "_attention")
                .c_str() /*name*/
        );
        break;
      }
      default: {
        assert(false);
      }
    }

    ff.residual_layer_norm(
        hidden_states,
        mha,
        nullptr,
        res_ln_outputs,
        false,
        axes,
        true,
        startcoder_config.layer_norm_epsilon,
        true,
        DT_NONE,
        std::string("layers_" + std::to_string(i) + "_ln_2").c_str());
    residual = res_ln_outputs[0];
    Tensor l2_norm = res_ln_outputs[1];

    // mlp
    Tensor c_fc = ff.dense(
        l2_norm,
        startcoder_config.intermediate_size,
        AC_MODE_NONE,
        true,
        DT_NONE,
        nullptr,
        nullptr,
        nullptr,
        REG_MODE_NONE,
        0.0f,
        std::string("layers_" + std::to_string(i) + "_mlp_c_fc").c_str());

    c_fc = ff.gelu(c_fc);

    c_proj = ff.dense(
        c_fc,
        startcoder_config.hidden_size,
        AC_MODE_NONE,
        true,
        DT_NONE,
        nullptr,
        nullptr,
        nullptr,
        REG_MODE_NONE,
        0.0f,
        std::string("layers_" + std::to_string(i) + "_mlp_c_proj").c_str());
  }
  // final normalization and linear
  ff.residual_layer_norm(residual,
                         c_proj,
                         nullptr,
                         res_ln_outputs,
                         false,
                         axes,
                         true,
                         startcoder_config.layer_norm_epsilon,
                         true,
                         DT_NONE,
                         "transformer_ln_f");
  Tensor ln_f = res_ln_outputs[1];

  Tensor lm_head = ff.dense(ln_f,
                            startcoder_config.vocab_size,
                            AC_MODE_NONE,
                            false,
                            DT_NONE,
                            nullptr,
                            nullptr,
                            nullptr,
                            REG_MODE_NONE,
                            0.0f,
                            "lm_head");

  Tensor output;
  if (mode == BEAM_SEARCH_MODE) {
    Tensor softmax = ff.softmax(lm_head, -1);
    // output = ff.beam_top_k(softmax, startcoder_config.max_beam_width, false);
    output = ff.argmax(softmax, /*beam_Search*/ true);
  } else {
    // Tensor softmax = ff.softmax(dense, -1);
    if (generationConfig.do_sample) {
      lm_head = ff.scalar_truediv(lm_head, generationConfig.temperature, false);
      Tensor softmax = ff.softmax(lm_head, -1);
      output = ff.sampling(softmax, generationConfig.topp);
    } else {
      // output = ff.arg_top_k(lm_head, /*k=*/1, false);
      output = ff.argmax(lm_head, /*beam_Search*/ false);
    }
  }

  InferenceManager *im = InferenceManager::get_inference_manager();
  FileDataLoader *fileloader = new FileDataLoader(
      "",
      weight_file_path,
      startcoder_config.num_attention_heads,
      1,
      startcoder_config.hidden_size,
      startcoder_config.hidden_size / startcoder_config.num_attention_heads,
      ff.config.tensor_parallelism_degree,
      use_full_precision);
  im->register_model_weights_loader(&ff, fileloader);
}

}; // namespace FlexFlow
